### DISCUSSION TITLE
If webview2 not installed, How can show message to my user?

### DISCUSSION THREAD

**Discussion by iloug** [tauri-apps/tauri#4867] [Q&A] [answered]
I need ship a portable exe version to my user, but on windows7, if user have not installed webview2, the exe just NOT run without any error message.

I want show a message just tell user need to install webview2, any ideas? thanks.

**Comment by FabianLars** [comment, marked as answer]
You need to check for registry keys (for example via the `winreg` crate) for example like this
```rs
if RegKey::predef(HKEY_LOCAL_MACHINE).open_subkey("SOFTWARE\\WOW6432Node\\Microsoft\\EdgeUpdate\\Clients\\{F3017226-FE2A-4295-8BDF-00C3A9A7E4C5}").is_err() // System-wide 64bit
  && RegKey::predef(HKEY_LOCAL_MACHINE).open_subkey("SOFTWARE\\Microsoft\\EdgeUpdate\\Clients\\{F3017226-FE2A-4295-8BDF-00C3A9A7E4C5}").is_err() // System-wide 32bit
  && RegKey::predef(HKEY_CURRENT_USER).open_subkey("SOFTWARE\\Microsoft\\EdgeUpdate\\Clients\\{F3017226-FE2A-4295-8BDF-00C3A9A7E4C5}").is_err() // User-wide 64bit&32bit
  {}
```
And to show a warning you'd need to use a dialog crate, like the one tauri uses: <https://github.com/PolyMeilex/rfd>.

Note that you could also download & install webview2 from within rust, for example like this
```rs
let installer = reqwest::blocking::get("https://go.microsoft.com/fwlink/p/?LinkId=2124703")?.bytes()?;
let path = std::env::temp_dir().join("MicrosoftEdgeWebview2Setup.exe");
if std::fs::write(&path, installer).is_ok() {
    let mut command = Command::new("powershell.exe");
    command.arg(path);
    command.arg("/install");
    let _ = dbg!(command.output()); // Do some error handling here instead of ignoring it.
}
```
Note that you need to use a rust library to download the installer instead of powershell for example or other OS tools, because windows 7 doesn't have TLS1.2 installed by default which is required by Microsoft's server.

**Reply by iloug** [reply]
Thanks a lot. I am newbie on rust, and will check crates winreg and rfd.

**Reply by thongnmtran** [reply]
Perfect, this is exactly what I'm looking for too. Thank you a lot.
